File: src/main/java/org/axostudio/axohologram/config/ConfigManager.java
package org.axostudio.axohologram.config;

import org.axostudio.axohologram.AxoHologram;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.configuration.file.YamlConfiguration;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.logging.Level;

public class ConfigManager {

    private final AxoHologram plugin;
    private FileConfiguration config;
    private FileConfiguration messages;
    private FileConfiguration media;
    private volatile VisibilityRuntimeConfig visibilityRuntimeConfig = VisibilityRuntimeConfig.from(null);
    private volatile DynamicRefreshRuntimeConfig dynamicRefreshRuntimeConfig = DynamicRefreshRuntimeConfig.from(null);

    private final File configFile;
    private final File messagesFile;
    private final File mediaFile;

    public ConfigManager(AxoHologram plugin) {
        this.plugin = plugin;
        this.configFile = new File(plugin.getDataFolder(), "config.yml");
        this.messagesFile = new File(plugin.getDataFolder(), "messages.yml");
        this.mediaFile = new File(plugin.getDataFolder(), "media.yml");
    }

    public void loadAllConfigs() {
        if (!plugin.getDataFolder().exists() && !plugin.getDataFolder().mkdirs()) {
            plugin.getLogger().warning("Could not create plugin data folder: " + plugin.getDataFolder().getAbsolutePath());
        }

        saveDefaultIfMissing("config.yml", configFile);
        saveDefaultIfMissing("messages.yml", messagesFile);
        saveDefaultIfMissing("media.yml", mediaFile);

        reloadConfig();
        reloadMessages();
        reloadMedia();
    }

    private void saveDefaultIfMissing(String resourcePath, File destination) {
        if (!destination.exists()) {
            plugin.saveResource(resourcePath, false);
        }
    }

    private FileConfiguration loadConfiguration(File file, String resourceName) {
        FileConfiguration configuration = YamlConfiguration.loadConfiguration(file);
        try (InputStream stream = plugin.getResource(resourceName)) {
            if (stream != null) {
                configuration.setDefaults(YamlConfiguration.loadConfiguration(new InputStreamReader(stream, StandardCharsets.UTF_8)));
                configuration.options().copyDefaults(true);
            }
        } catch (IOException exception) {
            plugin.getLogger().log(Level.SEVERE, "Could not load defaults for " + resourceName, exception);
        }
        return configuration;
    }

    public FileConfiguration getConfig() {
        if (config == null) {
            reloadConfig();
        }
        return config;
    }

    public FileConfiguration getMessages() {
        if (messages == null) {
            reloadMessages();
        }
        return messages;
    }

    public FileConfiguration getMedia() {
        if (media == null) {
            reloadMedia();
        }
        return media;
    }

    public void reloadConfig() {
        config = loadConfiguration(configFile, "config.yml");
        visibilityRuntimeConfig = VisibilityRuntimeConfig.from(config);
        dynamicRefreshRuntimeConfig = DynamicRefreshRuntimeConfig.from(config);
    }

    public void reloadMessages() {
        messages = loadConfiguration(messagesFile, "messages.yml");
    }

    public void reloadMedia() {
        media = loadConfiguration(mediaFile, "media.yml");
    }

    public void saveConfig() {
        saveYaml(config, configFile, "config");
    }

    public void saveMessages() {
        saveYaml(messages, messagesFile, "messages");
    }

    public void saveMedia() {
        saveYaml(media, mediaFile, "media");
    }

    public VisibilityRuntimeConfig getVisibilityRuntimeConfig() {
        return visibilityRuntimeConfig;
    }

    public DynamicRefreshRuntimeConfig getDynamicRefreshRuntimeConfig() {
        return dynamicRefreshRuntimeConfig;
    }

    private void saveYaml(FileConfiguration configuration, File file, String name) {
        if (configuration == null || !file.exists()) {
            return;
        }
        try {
            configuration.save(file);
        } catch (IOException ex) {
            plugin.getLogger().log(Level.SEVERE, "Could not save " + name + " to " + file, ex);
        }
    }
}

File: src/main/kotlin/org/axostudio/axohologram/config/ConfigManager.kt
package org.axostudio.axohologram.config

import org.axostudio.axohologram.common.text.MiniMessageUtil
import org.bukkit.configuration.file.FileConfiguration
import org.bukkit.configuration.file.YamlConfiguration
import org.bukkit.plugin.Plugin
import java.io.File
import java.io.InputStreamReader
import java.nio.charset.StandardCharsets

class ConfigManager(private val plugin: Plugin) {

    lateinit var configFile: File
    lateinit var animationsFile: File
    lateinit var mediaFile: File
    lateinit var visibilityFile: File
    lateinit var storageFile: File
    lateinit var langFolder: File

    var config: FileConfiguration = YamlConfiguration()
        private set
    var animationsConfig: FileConfiguration = YamlConfiguration()
        private set
    var mediaConfig: FileConfiguration = YamlConfiguration()
        private set
    var visibilityConfigFile: FileConfiguration = YamlConfiguration()
        private set
    var storageConfig: FileConfiguration = YamlConfiguration()
        private set
    var messages: FileConfiguration = YamlConfiguration()
        private set

    var pluginConfig: PluginConfig = PluginConfig()
        private set
    var runtimeConfig: RuntimeConfig = RuntimeConfig()
        private set
    var visibilityConfig: VisibilityConfig = VisibilityConfig()
        private set
    var animationConfig: AnimationConfig = AnimationConfig()
        private set

    fun loadAll() {
        if (!plugin.dataFolder.exists()) {
            plugin.dataFolder.mkdirs()
        }

        configFile = File(plugin.dataFolder, "config.yml")
        animationsFile = File(plugin.dataFolder, "animations.yml")
        mediaFile = File(plugin.dataFolder, "media.yml")
        visibilityFile = File(plugin.dataFolder, "visibility.yml")
        storageFile = File(plugin.dataFolder, "storage.yml")
        langFolder = File(plugin.dataFolder, "lang")

        if (!langFolder.exists()) {
            langFolder.mkdirs()
        }

        saveDefaultIfMissing("config.yml", configFile)
        saveDefaultIfMissing("animations.yml", animationsFile)
        saveDefaultIfMissing("media.yml", mediaFile)
        saveDefaultIfMissing("visibility.yml", visibilityFile)
        saveDefaultIfMissing("storage.yml", storageFile)
        val legacyMessagesFile = File(plugin.dataFolder, "messages.yml")
        val englishLanguageFile = findLanguageFile("en_US") ?: File(langFolder, "en_US.yml")
        val spanishLanguageFile = findLanguageFile("es_ES") ?: File(langFolder, "es_ES.yml")
        val hadLanguageFiles = hasLanguageFiles()

        saveDefaultIfMissing("lang/en_US.yml", englishLanguageFile)
        if (legacyMessagesFile.exists() && !hadLanguageFiles) {
            // Preserve the old Spanish/custom messages as the Spanish file,
            // then provide the bundled English file for the new format.
            runCatching { legacyMessagesFile.copyTo(spanishLanguageFile, overwrite = false) }
        } else {
            saveDefaultIfMissing("lang/es_ES.yml", spanishLanguageFile)
        }

        // A normal persisted hologram, provided once for a new installation.
        // Its marker prevents it from returning after an administrator deletes it.
        val exampleMarker = File(plugin.dataFolder, ".animation-example-installed")
        if (!exampleMarker.exists()) {
            saveDefaultIfMissing("holograms/animation_example.yml", File(plugin.dataFolder, "holograms/animation_example.yml"))
            exampleMarker.createNewFile()
        }

        reload()
    }

    fun reload() {
        config = loadYaml(configFile, "config.yml")
        animationsConfig = loadYaml(animationsFile, "animations.yml")
        mediaConfig = loadYaml(mediaFile, "media.yml")
        visibilityConfigFile = loadYaml(visibilityFile, "visibility.yml")
        storageConfig = loadYaml(storageFile, "storage.yml")

        pluginConfig = PluginConfig.from(config)
        runtimeConfig = RuntimeConfig.from(config)
        visibilityConfig = VisibilityConfig.from(if (visibilityFile.exists()) visibilityConfigFile else config)
        animationConfig = AnimationConfig.from(if (animationsFile.exists()) animationsConfig else config)

        val langFile = findLanguageFile(pluginConfig.language)
            ?: File(langFolder, "${pluginConfig.language}.yml")
        val fallbackLangFile = findLanguageFile("en_US")
            ?: File(langFolder, "en_US.yml")

        messages = if (langFile.exists()) {
            loadYaml(langFile, "lang/en_US.yml")
        } else if (fallbackLangFile.exists()) {
            loadYaml(fallbackLangFile, "lang/en_US.yml")
        } else {
            // Check legacy messages.yml
            val legacyMessages = File(plugin.dataFolder, "messages.yml")
            if (legacyMessages.exists()) loadYaml(legacyMessages, "lang/en_US.yml") else YamlConfiguration()
        }

        // Parsed messages are cached, so a language/prefix edit made before
        // /holo reload must invalidate the old components immediately.
        MiniMessageUtil.clearCache()
    }

    fun getMessage(key: String, def: String = key): String {
        val message = messages.getString(key) ?: def
        return if (key.equals("prefix", ignoreCase = true)) {
            message
        } else {
            withPrefix(message)
        }
    }

    fun withPrefix(message: String): String =
        MessageTemplateResolver.resolve(message, messages.getString("prefix").orEmpty())

    private fun saveDefaultIfMissing(resourcePath: String, destination: File) {
        if (!destination.exists()) {
            try {
                destination.parentFile?.mkdirs()
                plugin.saveResource(resourcePath, false)
            } catch (ignored: Exception) {
            }
        }
    }

    private fun hasLanguageFiles(): Boolean = langFolder.listFiles()
        ?.any { it.isFile && it.extension.equals("yml", ignoreCase = true) }
        ?: false

    private fun findLanguageFile(language: String): File? {
        val expectedName = "${language.substringBeforeLast(".yml", missingDelimiterValue = language)}.yml"
        return langFolder.listFiles()
            ?.firstOrNull { it.isFile && it.name.equals(expectedName, ignoreCase = true) }
    }

    private fun loadYaml(file: File, resourceName: String): FileConfiguration {
        val configuration = YamlConfiguration.loadConfiguration(file)
        plugin.getResource(resourceName)?.use { stream ->
            configuration.setDefaults(YamlConfiguration.loadConfiguration(InputStreamReader(stream, StandardCharsets.UTF_8)))
        }
        return configuration
    }
}
